File: src/cpp/multiprocessing/basic_process.cpp
/**
 * Basic Process Creation with fork()
 *
 * This example demonstrates the fundamental process creation mechanism
 * in Unix/Linux systems using fork().
 *
 * Key concepts:
 * - fork() creates a new process by duplicating the calling process
 * - Returns 0 in the child process, positive PID in the parent
 * - Child process has its own memory space (copy-on-write)
 * - waitpid() ensures parent waits for child to complete
 *
 * Unlike threads, processes have completely separate memory spaces
 * and require IPC (Inter-Process Communication) to share data.
 */

#include <iostream>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

void child_work() {
    std::cout << "Child process started\n";
    std::cout << "  Child PID: " << getpid() << "\n";
    std::cout << "  Parent PID: " << getppid() << "\n";
    std::cout << "Child process finished\n";
}

int main() {
    std::cout << "Main process started (PID: " << getpid() << ")\n";

    pid_t pid = fork();

    if (pid < 0) {
        std::cerr << "fork() failed\n";
        return 1;
    }

    if (pid == 0) {
        // Child process
        child_work();
    } else {
        // Parent process
        std::cout << "Parent waiting for child (PID: " << pid << ")...\n";
        waitpid(pid, nullptr, 0);
        std::cout << "Parent process finished\n";
    }

    return 0;
}

File: src/python/multiprocessing/basic_process.py
"""
Basic Process Example

This script demonstrates the fundamentals of multiprocessing in Python:
- Creating a child process using multiprocessing.Process
- Understanding the relationship between parent and child processes
- Accessing process and thread identification

Key Concepts:
- Each process has its own memory space (unlike threads which share memory)
- Child processes have different PIDs than the parent
- Each process has a main thread with its own thread ID
- Use join() to wait for child processes to complete

Unlike threads, processes bypass Python's GIL (Global Interpreter Lock),
making multiprocessing ideal for CPU-bound tasks.
"""

import os
import multiprocessing
import threading


def worker():
    """Worker function that runs in a child process."""
    print("Worker function")
    print(f"Worker function process id: {os.getpid()}")
    print(f"Worker thread id: {threading.current_thread().ident}")
    print(f"Worker thread name: {threading.current_thread().name}")


def main():
    print("Main function")
    print(f"Main function process id: {os.getpid()}")
    print(f"Main thread id: {threading.current_thread().ident}")
    print(f"Main thread name: {threading.current_thread().name}")

    # Create and start a child process
    process = multiprocessing.Process(target=worker)
    process.start()

    # Wait for the child process to complete
    process.join()

    print("Main function finished")


if __name__ == "__main__":
    main()
